### PR TITLE
Convert the message handler to async/await

### PR DESCRIPTION
**Pull request by clokep**
This is the last big piece of `synapse.handlers` to convert. It was a bit more involved since it is touch in so many places. I converted this method by method as separate commits since I had previously tried to convert this and failed.

The two interesting commits are 48c7f925edcf5c97e98235a1a7241c29c94decd6 and e76d407893509fe7e0f5469ce7bda0455508daa1. The rest are pretty mechanical.

### CODE REVIEW

File: synapse/handlers/message.py [hunk 1]
```diff
@@ -84,14 +84,22 @@ def __init__(self, hs):
...
             )
 
-    @defer.inlineCallbacks
-    def get_room_data(
-        self, user_id=None, room_id=None, event_type=None, state_key="", is_guest=False
-    ):
+    async def get_room_data(
+        self,
+        user_id: Optional[str] = None,
+        room_id: Optional[str] = None,
```
Comment (anoadragon453) [inline synapse/handlers/message.py:97]: Hmm, I know the previous docstring had it, but I'm not sure if `room_id` or `user_id` can be `None` here, after looking at `check_user_in_room_or_world_readable`'s docstring:

https://github.com/matrix-org/synapse/blob/38e1fac8861f12b707609da06008695a05aaf21c/synapse/api/auth.py#L638-L655

Feel free to look at/fix in a separate issue/PR though.

File: synapse/handlers/message.py [hunk 2]
```diff
@@ -454,31 +453,29 @@ def create_event(
...
+            txn_id
             prev_event_ids:
                 the forward extremities to use as the prev_events for the
                 new event.
 
                 If None, they will be requested from the database.
-
-            require_consent (bool): Whether to check if the requester has
+            require_consent: Whether to check if the requester has
                 consented to privacy policy.
```
Comment (anoadragon453) [inline synapse/handlers/message.py:464]:
```suggestion
                consented to the privacy policy.
```

File: synapse/handlers/message.py [hunk 3]
```diff
@@ -736,25 +735,25 @@ async def create_and_send_nonmember_event(
...
-    @defer.inlineCallbacks
-    def create_new_client_event(
-        self, builder, requester=None, prev_event_ids: Optional[Collection[str]] = None
-    ):
+    async def create_new_client_event(
+        self,
+        builder: EventBuilder,
+        requester: Optional[Requester] = None,
+        prev_event_ids: Optional[Collection[str]] = None,
+    ) -> Tuple[EventBase, EventContext]:
```
Comment (anoadragon453) [inline synapse/handlers/message.py:747]: I still see a number of `yield self.event_creation_handler.create_new_client_event` calls in the tests. Looks like `measure_func` saves us here, but I wondered whether you wanted to convert the `yield` calls anyways.

File: synapse/handlers/message.py [hunk 4]
```diff
@@ -736,25 +735,25 @@ async def create_and_send_nonmember_event(
...
+            requester:
             prev_event_ids:
                 the forward extremities to use as the prev_events for the
                 new event.
 
                 If None, they will be requested from the database.
 
         Returns:
-            Deferred[(synapse.events.EventBase, synapse.events.snapshot.EventContext)]
+            Tuple of created event, Context
```
Comment (anoadragon453) [inline synapse/handlers/message.py:760]:
```suggestion
            Tuple of created event, EventContext
```

or

```suggestion
            Tuple of created event, context
```

?

File: tests/test_visibility.py [hunk 5]
```diff
@@ -53,7 +53,7 @@ def test_filtering(self):
         #
 
         # before we do that, we persist some other events to act as state.
-        self.inject_visibility("@admin:hs", "joined")
+        yield self.inject_visibility("@admin:hs", "joined")
```
Comment (anoadragon453) [inline tests/test_visibility.py:56]: Use `self.get_success` instead so we don't need to update in future?

Comment (anoadragon453) [reply tests/test_visibility.py:56]: Though I do see this is consistent with other lines in this method...

Comment (clokep) [reply tests/test_visibility.py:56]: It is keeping it consistent with the rest of the test-case, yes.

File: synapse/handlers/message.py (hunk 1 above)
Comment (clokep) [reply synapse/handlers/message.py:97]: Yeah, looks incorrect to me too. I'll just fix it here, assuming mypy passes.

Comment (clokep) [reply synapse/handlers/message.py:97]: (There's also only 1 caller, which provides both parameters so...)

File: synapse/handlers/message.py (hunk 3 above)
Comment (clokep) [reply synapse/handlers/message.py:747]: I don't see these. There was a couple of instances that I wrapped in `defer.ensureDeferred`. Can you be more specific about where you see these?

**Comment by clokep** [comment]
@anoadragon453 Thanks for the review! I handled all your comments or responded, let me know if I should make more changes!

File: tests/test_visibility.py (hunk 5 above)
Comment (clokep) [reply tests/test_visibility.py:56]: (I think I said it last week, but updating all the tests is a separate task IMO.)

File: synapse/handlers/message.py (hunk 3 above)
Comment (anoadragon453) [reply synapse/handlers/message.py:747]: I don't see them either anymore. Must've been stale cache in PyCharm  ¯\\\_(ツ)_/¯

**Pull request merged**